plugins/font/plugin.js

```diff
@@ -88,131 +193,345 @@
+							this.setValue( value );
 						}
-					} else {
-						editor.removeStyle( previousStyle );
-					}
-				}
 
-				if ( value === this.defaultValue ) {
-					if ( previousStyle ) {
-						editor.removeStyle( previousStyle );
+						return;
 					}
-				} else {
-					editor.applyStyle( style );
-				}
 
-				editor.fire( 'saveSnapshot' );
-			},
+					// If no styles match, just empty it.
+					this.setValue( '', definition.defaultLabel );
+				}, this );
 
-			onRender: function() {
-				editor.on( 'selectionChange', function( ev ) {
-					var currentValue = this.getValue();
-
-					var elementPath = ev.data.path,
-						elements = elementPath.elements;
-
-					// For each element into the elements path.
-					for ( var i = 0, element; i < elements.length; i++ ) {
-						element = elements[ i ];
-
-						// Check if the element is removable by any of
-						// the styles.
-						for ( var value in styles ) {
-							if ( styles[ value ].checkElementMatch( element, true, editor ) ) {
-								if ( value != currentValue )
-									this.setValue( value );
-								return;
-							}
-						}
+				command.on( 'state', function() {
+					if ( command.state === CKEDITOR.TRISTATE_DISABLED ) {
+						this.setState( CKEDITOR.TRISTATE_DISABLED );
+					} else {
+						this.setState( CKEDITOR.TRISTATE_OFF );
 					}
-
-					// If no styles match, just empty it.
-					this.setValue( '', defaultLabel );
 				}, this );
 			},
 
 			refresh: function() {
-				if ( !editor.activeFilter.check( style ) )
+				if ( command.state === CKEDITOR.TRISTATE_DISABLED ) {
 					this.setState( CKEDITOR.TRISTATE_DISABLED );
+				}
 			}
 		} );
+
+		function getMatchingValue( editor, path, stylesData ) {
+			var elements = path.elements;
+
+			for ( var i = 0, element, value; i < elements.length; i++ ) {
+				element = elements[ i ];
+
+				// Check if the element is removable by any of
+				// the styles.
+				value = stylesData.findMatchingStyleName( function( style ) {
+					return style.checkElementMatch( element, true, editor );
+				}, this );
+
+				if ( value ) {
+					return value;
+				}
+			}
+
+			return;
+		}
+
+		function onClickHandler( newValue ) {
+			editor.focus();
+
+			var oldValue = this.getValue(),
+				newStyle = stylesData.getStyle( newValue ),
+				oldStyle = stylesData.getStyle( oldValue ),
+				range = editor.getSelection().getRanges()[ 0 ],
+				isRemoveOperation = newValue === removeStyleValue || newValue === undefined,
+				styleToRemove,
+				hasFragmentsWithoutNewStyle;
+
+			styleToRemove = isRemoveOperation && hasStyleToRemove( {
+				range: range,
+				configStyleDefinition: definition.configStyleDefinition
+			} );
+
+			hasFragmentsWithoutNewStyle = !isRemoveOperation && !hasAlreadyAppliedNewStyle( {
+				range: range,
+				style: newStyle,
+				configStyleDefinition: definition.configStyleDefinition
+			} );
+
+			if ( styleToRemove || hasFragmentsWithoutNewStyle ) {
+				editor.fire( 'saveSnapshot' );
```

Comment:
The idea is the same as in the above comment. Firstly it is determined if style should be applied or not. So it is analyzed selection in the editor. This is nonintrusive check and doesn't sound to require `saveSnapshot`. The snapshot is taken before the command is going to be executed, so when is decision to change something in the editor's content.
From my perspective in this flow, it is not required to save a snapshot at the beginning. There might be a situation when the font command won't be executed, so firing snapshot is not necessary in such cases. Do you find some idea when it might be useful?